src/types/utils.cpp

```diff
@@ -99,6 +100,219 @@ til::color Utils::ColorFromHexString(const std::string_view str)
     return til::color{ r, g, b };
 }
 
+// Routine Description:
+// - Given a color spec string, attempts to parse the color that's encoded.
+//   Based on the XParseColor documentation, the supported specs currently are the following:
+//      spec1: a color in the following format:
+//          "rgb:<red>/<green>/<blue>"
+//      spec2: a color in the following format:
+//          "#<red><green><blue>"
+//      spec3: The XOrg app color names:
+//          "orange"
+//
+//   In the first two specs, <color> is a value contains up to 4 hex digits, upper or lower case.
+// Arguments:
+// - string - The string containing the color spec string to parse.
+// Return Value:
+// - An optional color which contains value if a color was successfully parsed
+std::optional<til::color> Utils::ColorForXParseColorSpec(const std::wstring_view string) noexcept
+try
+{
+    bool foundXParseColorSpec = false;
+    bool foundValidColorSpec = false;
+
+    bool isSharpSignFormat = false;
+    size_t rgbHexDigitCount = 0;
+    std::array<unsigned int, 3> colorValues = { 0 };
+    std::array<unsigned int, 3> parameterValues = { 0 };
+    const auto stringSize = string.size();
+
+    // First we look for "rgb:"
+    // Other colorspaces are theoretically possible, but we don't support them.
+    auto curr = string.cbegin();
+    if (stringSize > 4)
+    {
+        auto prefix = std::wstring(string.substr(0, 4));
+        std::transform(prefix.begin(), prefix.end(), prefix.begin(), ::towlower);
+        if (prefix.compare(L"rgb:") == 0)
+        {
+            // If all the components have the same digit count, we can have one of the following formats:
+            // 9 "rgb:h/h/h"
+            // 12 "rgb:hh/hh/hh"
+            // 15 "rgb:hhh/hhh/hhh"
+            // 18 "rgb:hhhh/hhhh/hhhh"
+            // Note that the component sizes aren't required to be the same.
+            // Anything in between is also valid, e.g. "rgb:h/hh/h" and "rgb:h/hh/hhh".
+            // Any fewer cannot be valid, and any more will be too many. Return early in this case.
+            if (stringSize < 9 || stringSize > 18)
+            {
+                return std::nullopt;
+            }
+
+            foundXParseColorSpec = true;
+
+            std::advance(curr, 4);
+        }
+    }
+
+    // Try the sharp sign format.
+    if (!foundXParseColorSpec && stringSize > 1)
+    {
+        const auto prefix = string.substr(0, 1);

```

Comment:
We don’t even need to use substr for this now :)
(Since we checked that the size is >1 and we’re using a single character check below)

Reply:
When I was writing this I couldn't stop thinking that how on earth `startswith` is a C++20 thing. The feeling just kepp getting stronger..